squidpy/gr/_sepal.py

```diff
@@ -0,0 +1,341 @@
+        else:
+            score.append(np.nan)
+
+        if queue is not None:
+            queue.put(Signal.UPDATE)
+
+    if queue is not None:
+        queue.put(Signal.FINISH)
+
+    return np.array(score)
+
+
+@njit(parallel=False, fastmath=True)
+def _diffusion(
+    conc: np.ndarray,
+    laplacian: Callable[..., np.float_],
+    n_iter: int,
+    sat: np.ndarray,
+    sat_idx: np.ndarray,
+    sat_shape: int,
+    unsat: np.ndarray,
+    unsat_idx: np.ndarray,
+    conc_shape: int,
+    dt: float = 0.001,
+    D: float = 1.0,
+    thres: float = 1e-8,
+) -> np.ndarray:
+    """Simulate diffusion process on a regular graph."""
+    entropy_arr = np.zeros(n_iter)
+    ent = np.ones(n_iter + 1)
+
+    nhood = np.zeros(sat_shape)
+    weights = np.ones(sat_shape)
+
+    for i in range(n_iter):
+        for j in range(sat_shape):
+            nhood[j] = np.sum(conc[sat_idx[j]])
+        d2 = laplacian(conc[sat], nhood, weights)
+
+        dcdt = np.zeros(conc_shape)
+        dcdt[sat] = D * d2
+        conc[sat] += dcdt[sat] * dt
+        conc[unsat] += dcdt[unsat_idx] * dt
+        # set values below zero to 0
+        conc[conc < 0] = 0
+        # compute entropy
+        ent[i + 1] = _entropy(conc[sat]) / sat_shape
+        entropy_arr[i] = np.abs(ent[i + 1] - ent[i])  # estimate entropy difference
+        if entropy_arr[i] <= thres:
+            break
+
+    return np.nonzero(entropy_arr <= thres)[0]
+
+
+# taken from https://github.com/almaan/sepal/blob/master/sepal/models.py
+@njit(parallel=False, fastmath=True)
+def _laplacian_rect(
+    centers: np.ndarray,
+    nbrs: np.ndarray,
+    h: np.ndarray,
+) -> np.float_:
+    """Laplacian approx rectilinear grid."""
+    d2f = nbrs - 4 * centers
+    d2f = d2f / h ** 2
+
+    return d2f
+
+
+# taken from https://github.com/almaan/sepal/blob/master/sepal/models.py
+@njit(parallel=False, fastmath=True)
+def _laplacian_hex(
+    centers: np.ndarray,
+    nbrs: np.ndarray,
+    h: np.ndarray,
+) -> np.float_:
+    """Laplacian approx hexagonal grid."""
+    d2f = nbrs - 6 * centers
+    d2f = d2f / h ** 2
+    d2f = (d2f * 2) / 3
+
+    return d2f
+
+
+# taken from https://github.com/almaan/sepal/blob/master/sepal/models.py
+@njit(parallel=False, fastmath=True)
+def _entropy(
+    xx: np.ndarray,
+) -> np.float_:
+    """Entropy of array."""
+    xnz = xx[xx > 0]
+    xs = np.sum(xnz)
+    xn = xnz / xs
+    xl = np.log(xn)
+    return (-xl * xn).sum()
+
+
+def _compute_idxs(
+    g: spmatrix, spatial: np.ndarray, sat_thres: int, metric: str = "l1"
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """Get saturated and unsaturated nodes and nhood indexes."""
+    sat, unsat = _get_sat_unsat_idx(g.indptr, g.shape[0], sat_thres)
+
+    sat_idx, nearest_sat, un_unsat = _get_nhood_idx(sat, unsat, g.indptr, g.indices, sat_thres)
+
+    # compute dist btwn remaining unsat and all sat
+    dist = pairwise_distances(spatial[un_unsat], spatial[sat], metric=metric)
```

Comment:
Maybe `force_all_finite=...` from https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html
could be set to something other than default (e.g. allowing NaN and/but no infinities + use, np.nanargmin later).
Just a thought, not sure if makes sense.

Reply:
I see what you mean but if there is np.nan/inf in arr then we'd not be able to assign it to closest saturated neighbor so would not work I think

Reply:
Ok, I see. Let's keep it as it is.